Add get and publish images to client

jessjenkins opened the pull request:
Add image api client code.  It implements the following endpoints of the [dp-image-api](https://github.com/ONSdigital/dp-image-api)…
- Get a list of images belonging to a collection `GET /images`
- Publish an image by using `POST /images/{id}/publish`

This client is based on the [dp-dataset-api-java-client](https://github.com/ONSdigital/dp-dataset-api-java-client).

jessjenkins committed:
Add image api client

jessjenkins committed:
Add unit test

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+        this.serviceAuthToken = serviceAuthToken;
+    }
+
+    public ImageAPIClient(String imageAPIURL, String serviceAuthToken) throws URISyntaxException {
```
Missing Javadoc.

daiLlew replied:
Need to added the param description to the javadoc - copy and past from other constructor.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+     * @return
+     * @throws IOException
+     * @throws ImageAPIException
+     */
```
Incomplete javadoc.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+    @Override
+    public Images getImagesWithCollectionId(String collectionID) throws IOException, ImageAPIException {
+
+        String path = "/images?collection_id=" + collectionID;
```
Should we be validating the collection ID is not null or empty? Saves an unnecessary HTTP request.

jessjenkins replied:
I've added a validation method a little like the publish method has. I had considered making the collectionID optional and just having a `getImages` method but this allows all images to be returned which could be undesirable. I'm open to challenge on this.

daiLlew replied:
Good point. I guess the Image API should determine if this validation is necessary i.e. if the API current expects the collection ID param then I feel the validation here is justified, if its an optional param then it's not.

jessjenkins replied:
The API currently allows the param to be optional, but if I were to do that here, I'd probably want to change the method name to just `getImages()`. I think that the client method `getImagesWithCollectionID()` would acceptably expect a collectionID to be supplied. I don't mind either approach but the `getImages()` approach is probably more correct for a more generic client. I am minded to make the changes to the name and make the param optional (excluded on nil/"").

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+                    throw new UnauthorisedException();
+                default:
+                    throw new UnexpectedResponseException(
+                            formatErrResponse(req, resp), resp.getStatusLine().getStatusCode());
```
use the `statusCode` var - its shorter 😂

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+                    throw new UnauthorisedException();
+                default:
+                    throw new UnexpectedResponseException(
+                            formatErrResponse(req, resp), resp.getStatusLine().getStatusCode());
```
Use the `statusCode` var

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private static boolean isNotEmpty(String str) {
+        return str != null && str.length() > 0;
```
You can remove the need for this method and use `StringUtils.IsNotEmpty`.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/RetryStrategy.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public RetryStrategy() {
+        this(3, 20);
```
`(3, 20)` ==  [Magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))

Are these the defaults? If so consider making them constants with meaningful names. Makes it clear and obvious what they are and what purpose they serve.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/RetryStrategy.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public boolean retryRequest(HttpResponse response, int executionCount, HttpContext context) {
+        return executionCount <= maxRetries &&
+                response.getStatusLine().getStatusCode() >= HttpStatus.SC_INTERNAL_SERVER_ERROR;
```
Just checking the logic: Are we only retrying if its a 500 or above?

jessjenkins replied:
This is a very good question. This class is pretty much a copy of [the equivalent in the dataset api client](https://github.com/ONSdigital/dp-dataset-api-java-client/blob/master/src/main/java/dp/api/dataset/RetryStrategy.java) and I haven't given it much thought. I'll consider this with a tech lead.

jessjenkins replied:
Having discussed this with the tech leads we're going to keep the functionality for now consistent with the dataset client and will revisit the subject of retries more holistically as a separate piece of work.

daiLlew replied:
👍 I did see the conversation. I asked the question because I wanted to clarify that was the intended behaviour. If that is what has been decided then I will resolve this comment.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/exception/BadRequestException.java:
```diff
@@ -0,0 +1,8 @@
+    public BadRequestException(String message) {
+        super(message);
+    }
+}
```
I'm not sure I see the value in having separate implementations of `ImageAPIException` as they do not contain any functionality in them that is specific to error they are handling. For example: The only difference between `ForbiddenException` & `UnauthorisedException` is the name.

Consider removing these subclasses and passing the error context as a parameter to the `ImageAPIException` constructor. 

If you need different functionality/message for error each error type consider writing static methods in the Exception class that construct the Exception with the details you need. e.g.:

````
public class ImageAPIException extends Exception {

    public ImageAPIException(String message) {
        // insert logic here.
    }

    public static ImageAPIException forbiddenException(req, resp) {
        return new ImageAPIException(// format the error message with the appropriate details //)
    }

    public static ImageAPIException badRequestException(req, resp) {
        return new ImageAPIException(// format the error message with the appropriate details //)
    }
    .... etc etc.
}
````

jessjenkins replied:
I've refactored this to a single exception class.

daiLlew commented on src/test/java/com/github/onsdigital/dp/image/api/client/RetryStrategyTest.java:
```diff
@@ -0,0 +1,69 @@
+    private final int maxRetries = 3;
+    private final int retryInterval = 20;
+
+    private final RetryStrategy retryStrategy = new RetryStrategy(maxRetries, retryInterval);
```
Be aware this same object and it's state will be shared by all test cases. I think this is what you are expecting however this does mean the outcome of each test is dependant on the previous tests running/completing. It could lead to some confusing/misleading test failures. I don't think this is a problem here as this class is probably never going to change but in general test cases should be autonomous.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/model/State.java:
```diff
@@ -0,0 +1,18 @@
+    @JsonProperty("created")
+    CREATED,
+
+    @JsonProperty("edition-confirmed")
```
I might be wrong but I think our JSON standards state we should use underscore i.e. `edition_confirmed`. Ignore me if that is not the case.

jessjenkins replied:
This file shouldn't exist. It got copy-pasted from the dataset api client and isn't relevant here. Deleting it.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageClient.java:
```diff
@@ -0,0 +1,21 @@
+import java.io.Closeable;
+import java.io.IOException;
+
+public interface ImageClient extends Closeable {
```
Javadoc

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,149 @@
+
+import static com.github.onsdigital.logging.v2.event.SimpleEvent.info;
+
+public class ImageAPIClient implements ImageClient {
```
Class Javadoc.

jessjenkins committed:
Remove accidentally copied unrelated model class

jessjenkins committed:
Adress PR comments

- Add JavaDoc
- Add constants for hardcoded values
- Use StringUtils.isNotEmpty instead of custom method
- Add validation for empty collectionId on getImagesWithCollectionI

jessjenkins committed:
Refactor ImageAPI exceptions to single class

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/ImageAPIClient.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    private String formatErrResponse(HttpRequestBase httpRequest, CloseableHttpResponse response) {
+        return String.format("the image api returned a %s response for %s",
+                response.getStatusLine().getStatusCode(),
+                httpRequest.getURI());
```
Is there value in including the expected status code as well? My thinking is that it will provide instant clarity as to the expected behaviour without having to consult the code.

daiLlew commented on src/main/java/com/github/onsdigital/dp/image/api/client/exception/ImageAPIException.java:
```diff
@@ -0,0 +1,20 @@
+public class ImageAPIException extends Exception {
+    private int code;
+
+    public ImageAPIException(String message, int code) {
```
Java doc - `message` is obvious but specifying what `code` is will be useful to anyone using this class.

jessjenkins committed:
Add expected code to ex'n msg and update JavaDoc

jessjenkins committed:
Collection ID is optional on getImages

daiLlew reviewed (approved):
👍
